File: packages/public-sdk-js/src/resources/deploys.ts
import type { RequestInit } from 'node-fetch';
import { randomUUID } from 'node:crypto';
import type { JsonFetcher } from './client.js';

const generateIdempotencyKey = () => {
  if (typeof globalThis.crypto?.randomUUID === 'function') {
    return globalThis.crypto.randomUUID();
  }
  return randomUUID();
};

export interface CreateDeployInput {
  serviceId: string;
  environment: 'staging' | 'production';
  gitRef: string;
  metadata?: Record<string, unknown>;
}

export interface Deploy {
  releaseId: string;
  status: 'pending' | 'ready' | 'promoting' | 'promoted';
  auditId?: string;
}

export class DeploysResource {
  constructor(private readonly fetcher: JsonFetcher) {}

  create(input: CreateDeployInput): Promise<Deploy> {
    const init: RequestInit = {
      method: 'POST',
      headers: { 'Idempotency-Key': generateIdempotencyKey() },
      body: JSON.stringify(input),
    };
    return this.fetcher('/v1/deploys', init) as Promise<Deploy>;
  }

  promote(releaseId: string): Promise<{ releaseId: string; status: 'promoting' | 'promoted' }> {
    return this.fetcher(`/v1/releases/${releaseId}:promote`, { method: 'POST' }) as Promise<{
      releaseId: string;
      status: 'promoting' | 'promoted';
    }>;
  }
}

File: packages/public-sdk-py/blackroad/deploys.py
"""Deploy resource helpers."""
from __future__ import annotations

from typing import Any, Awaitable, Callable, Dict
import uuid

RequestFn = Callable[[str, Dict[str, Any]], Awaitable[Any]]


class DeploysResource:
    """Interface for deploy endpoints."""

    def __init__(self, requester: RequestFn) -> None:
        self._request = requester

    async def create(self, payload: Dict[str, Any]) -> Any:
        headers = {"Idempotency-Key": str(uuid.uuid4())}
        data = await self._request("/v1/deploys", {"method": "POST", "json": payload, "headers": headers})
        return data

    async def promote(self, release_id: str) -> Any:
        return await self._request(f"/v1/releases/{release_id}:promote", {"method": "POST"})
